Source/Core/Common/OnionConfig.h

```diff
@@ -0,0 +1,152 @@
+// Copyright 2010 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <list>
+#include <map>
+#include <memory>
+#include <string>
+
+// XXX: Purely for case insensitive compare
+#include "Common/IniFile.h"
+
+namespace OnionConfig
+{
+	enum class OnionLayerType
+	{
+		LAYER_BASE,
+		LAYER_GAME,
+		LAYER_NETPLAY,
+		LAYER_MOVIE,
+		LAYER_COMMANDLINE,
+		LAYER_CURRENTRUN,
+		LAYER_META,
+	};
+
+	enum class OnionSystem
+	{
+		SYSTEM_MAIN,
+		SYSTEM_GCPAD,
+		SYSTEM_WIIPAD,
+		SYSTEM_GFX,
+		SYSTEM_LOGGER,
+		SYSTEM_DEBUGGER,
+		SYSTEM_UI,
+	};
+
+	class OnionPetal;
+	class BloomLayer;
+	class ConfigLayerLoader;
+
+	// XXX: Make unordered
+	using PetalValueMap = std::map<std::string, std::string, CaseInsensitiveStringCompare>;
+	using BloomLayerMap = std::map<OnionSystem, std::list<OnionPetal*>>;
+	// Has to be ordered!
+	using OnionBloom = std::map<OnionLayerType, BloomLayer*>;
+
+	class OnionPetal
+	{
+		friend BloomLayer;
+		friend ConfigLayerLoader;
+
+	public:
+		OnionPetal(OnionLayerType layer, OnionSystem system, const std::string& name)
+			: m_layer(layer), m_system(system), m_name(name) {}
+
+		virtual bool Exists(const std::string& key) const;
+		bool Delete(const std::string& key);
+
+		// Setters
+		virtual void Set(const std::string& key, const std::string& value);
+
+		void Set(const std::string& key, u32 newValue);
+		void Set(const std::string& key, float newValue);
+		void Set(const std::string& key, double newValue);
+		void Set(const std::string& key, int newValue);
+		void Set(const std::string& key, bool newValue);
+		void Set(const std::string& key, const std::vector<std::string>& newValues);
```

Comment:
Is there a way to use templates? Feel free to pick https://github.com/degasus/dolphin/commit/6dda75af7ce4eb1ca2444d78f228444529d9c96b